great_expectations/expectations/metrics/util.py

```diff
@@ -1058,3 +1060,165 @@ def get_sqlalchemy_source_table_and_schema(
+        expectation_domain_column_name: column that Expectation is being run for
+        unexpected_index_column_names: column_names for indices, either named index or unexpected_index_columns
+
+    Returns:
+        List of Dicts that contain ID/PK values
+    """
+    if expectation_domain_column_name is None:
+        raise ge_exceptions.MetricResolutionError(
+            message=f"Error: The domain column is currently set to None. Please check your configuration.",
+            failed_metrics=["unexpected_index_list"],
+        )
+
+    domain_records_df_index_names: List[str] = domain_records_df.index.names
+    unexpected_indices: List[Union[int, str]] = list(domain_records_df.index)
+
+    for column_name in unexpected_index_column_names:
+        if column_name not in domain_records_df_index_names:
+            raise ge_exceptions.MetricResolutionError(
+                message=f"Error: The column {column_name} does not exist in the named indices. Please check your configuration",
+                failed_metrics=["unexpected_index_list"],
+            )
+
+    unexpected_index_list: List[Dict[str, Any]] = list()
+
+    for index in unexpected_indices:
+        primary_key_dict: Dict[str, Any] = dict()
+        # domain column first
+        primary_key_dict[expectation_domain_column_name] = domain_records_df.at[
+            index, expectation_domain_column_name
+        ]
+        for column_name in unexpected_index_column_names:
+            tuple_index = domain_records_df_index_names.index(column_name, 0)
+            primary_key_dict[column_name] = index[tuple_index]
+        unexpected_index_list.append(primary_key_dict)
+    return unexpected_index_list
+
+
+def get_unexpected_indices_for_single_pandas_named_index(
+    domain_records_df: pd.DataFrame,
+    unexpected_index_column_names: List[str],
+    expectation_domain_column_name: Union[str, None] = None,
+) -> List[Dict[str, Any]]:
+    """
+    Builds unexpected_index list for Pandas Dataframe in situation where the named
+    columns is also a named index. This method handles the case when there is a single named index.
+    Args:
+        domain_records_df: reference to Pandas dataframe
+        expectation_domain_column_name: column that Expectation is being run on.
+        unexpected_index_column_names: column_names for indices, either named index or unexpected_index_columns
+
+    Returns:
+        List of Dicts that contain ID/PK values
+
+    """
+    unexpected_index_values_by_named_index: List[Union[int, str]] = list(
+        domain_records_df.index
+    )
+    unexpected_index_list: List[Dict[str, Any]] = list()
+
+    if not (
+        len(unexpected_index_column_names) == 1
+        and unexpected_index_column_names[0] == domain_records_df.index.name
+    ):
+        raise ge_exceptions.MetricResolutionError(
+            message=f"Error: The column {unexpected_index_column_names[0]} does not exist in the named indices. Please check your configuration",
```

Comment:
Calling `unexpected_index_column_names[0]` in the error message might die if `unexpected_index_column_names[0]` is empty. Maybe have another `if` block above this one that checks `if not unexpected_index_column_names: raise ge_exceptions.MetricResolutionError("unexpected_index_column_names must not be empty")`

Reply:
Another simpler solution would be to change `{unexpected_index_column_names[0]}` to `{unexpected_index_column_names[0] if unexpected_index_column_names else "<no column specified>"}` or something similar.